aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,907 @@
+from datetime import datetime, timedelta
+import sys
+from typing import Any, Iterator, Sequence, TypeVar, Iterable
+
+import asyncio
+import aiohttp
+import pandas as pd
+import google.cloud.bigquery as bq
+
+from google.cloud import secretmanager
+from google.api_core.exceptions import ClientError
+
+
+logger = logging.getLogger('Cost Aggregate')
+
+handler = logging.StreamHandler(sys.stderr)
+handler.setFormatter(logging.Formatter('%(asctime)s %(levelname)s: %(message)s'))
+
+logger.addHandler(handler)
+if os.getenv('DEV') in ('1', 'true', 'yes'):
+    logger.setLevel(logging.DEBUG)
+
+# pylint: disable=invalid-name
+T = TypeVar('T')
+
+DEFAULT_TOPIC = 'admin'
+
+GCP_PROJECT = 'billing-admin-290403'
+
+GCP_BILLING_BQ_TABLE = (
+    f'{GCP_PROJECT}.billing.gcp_billing_export_v1_01D012_20A6A2_CBD343'
+)
+GCP_AGGREGATE_DEST_TABLE = os.getenv(
+    'GCP_AGGREGATE_DEST_TABLE', 'billing-admin-290403.billing_aggregate.aggregate'
+)
+
+assert GCP_AGGREGATE_DEST_TABLE
+logger.info(f'GCP_AGGREGATE_DEST_TABLE: {GCP_AGGREGATE_DEST_TABLE}')
+
+IS_PRODUCTION = os.getenv('PRODUCTION') in ('1', 'true', 'yes')
+
+# 10% overhead
+HAIL_SERVICE_FEE = 0.05
+DEFAULT_BQ_CHUNK_SIZE = 9000
+ANALYSIS_RUNNER_PROJECT_ID = 'analysis-runner'
+DEFAULT_RANGE_INTERVAL = timedelta(days=2)
+
+
+HAIL_BASE = 'https://batch.hail.populationgenomics.org.au'
+HAIL_UI_URL = HAIL_BASE + '/batches/{batch_id}'
+HAIL_BATCHES_API = HAIL_BASE + '/api/v1alpha/batches'
+HAIL_JOBS_API = HAIL_BASE + '/api/v1alpha/batches/{batch_id}/jobs/resources'
+
+HAIL_PROJECT_FIELD = {
+    'id': 'hail-295901',
+    'number': '805950571114',
+    'name': 'hail-295901',
+    'labels': [],
+    'ancestry_numbers': '/648561325637/',
+    'ancestors': [
+        {
+            'resource_name': 'projects/805950571114',
+            'display_name': 'hail-295901',
+        },
+        {
+            'resource_name': 'organizations/648561325637',
+            'display_name': 'populationgenomics.org.au',
+        },
+    ],
+}
+
+SEQR_PROJECT_FIELD = {
+    'id': 'seqr-308602',
+    'number': '1021400127367',
+    'name': 'seqr-308602',
+    'labels': [],
+    'ancestry_numbers': '/648561325637/',
+    'ancestors': [
+        {
+            'resource_name': 'organizations/648561325637',
+            'display_name': 'populationgenomics.org.au',
+        }
+    ],
+}
+
+
+_BQ_CLIENT: bq.Client = None
+
+
+def get_bigquery_client():
+    """Get instantiated cached bq client"""
+    global _BQ_CLIENT
+    if not _BQ_CLIENT:
+        _BQ_CLIENT = bq.Client()
+    return _BQ_CLIENT
+
+
+async def async_retry_transient_get_json_request(
+    url,
+    errors: Exception | tuple[Exception, ...],
+    *args,
+    attempts=5,
+    session=None,
+    **kwargs,
+):
+    """
+    Retry a function with exponential backoff.
+    """
+
+    async def inner_block(session):
+        for attempt in range(1, attempts + 1):
+            try:
+                async with session.get(url, *args, **kwargs) as resp:
```

Comment:
Is there an implicit timeout for this call? Apart from an exponential backoff, good standard practice is to have timeouts for all requests.

Reply:
There is an implicit timeout for aiohttp - though they [include the queue time](https://github.com/aio-libs/aiohttp/issues/3203) (if n_requests > size of pool). Do you have suggestions for an explicit timeout, otherwise is it okay to leave it implicit for this case?

Reply:
I've added an explicit 60 second timeout here.